contracts/UniversalStorageProof_SpokePool.sol

```diff
@@ -0,0 +1,159 @@
+        address _hubPoolStore,
+        address _wrappedNativeTokenAddress,
+        uint32 _depositQuoteTimeBuffer,
+        uint32 _fillDeadlineBuffer,
+        IERC20 _l2Usdc,
+        ITokenMessenger _cctpTokenMessenger
+    )
+        SpokePool(_wrappedNativeTokenAddress, _depositQuoteTimeBuffer, _fillDeadlineBuffer)
+        CircleCCTPAdapter(_l2Usdc, _cctpTokenMessenger, CircleDomainIds.Ethereum)
+    {
+        helios = _helios;
+        hubPoolStore = _hubPoolStore;
+    }
+
+    function initialize(
+        uint32 _initialDepositId,
+        address _crossDomainAdmin,
+        address _withdrawalRecipient
+    ) public initializer {
+        __SpokePool_init(_initialDepositId, _crossDomainAdmin, _withdrawalRecipient);
+    }
+
+    /**
+     * @notice This can be called by an EOA to relay call data stored on the HubPool on L1 into this contract.
+     * @dev Consider making this an onlyOwner function so that only a privileged EOA can relay state. This EOA
+     * wouldn't be able to tamper with the _publicValues but we can reduce the chance of replay-attacks this way if
+     * we set the EOA to a trusted actor. Replay attacks are possible if this contract has the same address
+     * on multiple chains.
+     * @param _slotKey Slot storage hash
+     * @param _value Slot storage value
+     * @param _blockNumber Block number in light client we want to check slot value of slot key
+     */
+    function receiveL1State(
+        bytes32 _slotKey,
+        bytes calldata _value,
+        uint256 _blockNumber
+    ) external validateInternalCalls {
+        bytes32 expectedSlotValueHash = keccak256(_value);
+
+        // Verify Helios light client is aware of the storage slot:
+        bytes32 slotValueHash = IHelios(helios).getStorageSlot(_blockNumber, hubPoolStore, _slotKey);
+        if (expectedSlotValueHash != slotValueHash) {
+            revert SlotValueMismatch();
+        }
+
+        // Validate state is intended to be sent to this contract. The target could have been set to the zero address
+        // which is used by the StorageProof_Adapter to denote messages that can be sent to any target.
+        (address target, bytes memory message) = abi.decode(_value, (address, bytes));
+        if (target != address(0) && target != address(this)) {
+            revert NotTarget();
+        }
+
+        // Prevent replay attacks.
+        bytes32 dataHash = _slotKey;
+        if (verifiedProofs[dataHash]) {
+            revert AlreadyReceived();
+        }
+        verifiedProofs[dataHash] = true;
+        emit VerifiedProof(dataHash, msg.sender);
+
+        _executeMessage(message);
+    }
+
+    /**
+     * @notice This function is only callable by the owner and is used as an emergency fallback to relay
+     * a message to this SpokePool in the case where the light-client is not functioning correctly.
+     */
+    function adminReceiveL1State(bytes memory message) external onlyOwner validateInternalCalls {
```

Comment:
Open question: do we need this fallback function to circumvent the light client (in case it breaks for some reason), and if so, is this is the right implementation?

Reply:
I think we do. I think the open question is who the owner is and if there should be any verification about the light client being up to date.

Reply:
> if there should be any verification about the light client being up to date (i.e. should we not allow the owner to override unless the light client is < n blocks behind the current timestamp).

WDYT about a check like `if (block.number - BUFFER <= helios.head()) revert AdminCallTooCloseToHeliosHead` ?

Reply:
Yeah, something like that would make sense. But we need to relate block.timestamp to helios slots -- I'll check on if/how we can do this. We can't use the local block.number value since that's the block number on the local chain, not mainnet. The block numbers in helios are mainnet block numbers.

Reply:
Helios PR should be updated to provide the `headTimestamp()` function.